lib/internal/modules/cjs/loader.js

```diff
@@ -937,6 +939,28 @@ Module._load = function(request, parent, isMain) {
   const mod = loadNativeModule(filename, request);
   if (mod && mod.canBeRequiredByUsers) return mod.exports;
 
+  // Only return the file location if it is in an allowed location
+  // or the allowed locations are not set
+  let locationAllowed = true;
+  if (allowedModuleLocations && allowedModuleLocations.length > 0) {
+    locationAllowed = false;
+    debug('Allowed module locations count: %d', allowedModuleLocations.length);
+    for (const allowedLocation of allowedModuleLocations) {
+      const relative = path.relative(allowedLocation, filename);
+      debug('Relative path from allowed location "%s" to loaded "%s": %s',
+            allowedLocation, filename, relative);
+      if (relative === '' ||
+          (!relative.startsWith('..') && !path.isAbsolute(relative))) {
+        debug('Module will be allowed');
+        locationAllowed = true;
+        break;
+      }
+    }
+  }
```

Comment:
should cache builtins to avoid mutations from affecting this (comment 2 of n)

```suggestion
  if (allowedModuleLocations && allowedModuleLocations.length > 0) {
    locationAllowed = false;
    debug('Allowed module locations count: %d', allowedModuleLocations.length);
    // avoids mutation of Array.prototype[Symbol.iterator] from affecting this
    for (let i = 0; i < allowedModuleLocations.length; i++) {
      const allowedLocation = allowedModuleLocations[i];
      const relative = PathRelative(allowedLocation, filename);
      debug('Relative path from allowed location "%s" to loaded "%s": %s',
            allowedLocation, filename, relative);
      if (relative === '' ||
          (!StringPrototypeStartsWith(relative, '..') && !PathIsAbsolute(relative))) {
        debug('Module will be allowed');
        locationAllowed = true;
        break;
      }
    }
  }
```